cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataMigratorTest.java

```diff
@@ -311,11 +290,100 @@ private Put createValuePut(NamespacedEntityId targetId, String key, String value
   private Put createHistoryPut(NamespacedEntityId targetId, long time) {
     MetadataV1 metadataV1 = getMetadataV1(targetId);
     byte[] row = MdsHistoryKey.getMdsKey(targetId, time).getKey();
+    String jsonValue = getJsonValue(targetId.getEntityType());
+
     Put put = new Put(row);
-    put.add(Bytes.toBytes("h"), Bytes.toBytes(GSON.toJson(metadataV1)));
+    put.add(Bytes.toBytes("h"), Bytes.toBytes(jsonValue));
     return put;
   }
 
+  private String getJsonValue(EntityType entityType) {
+    String value = "";
+    switch (entityType) {
+      case APPLICATION:
+        value = "{\n" +
```

Comment:
Never construct a json string using string concatenation. It is super difficult to understand and maintain. Please create object and use Gson to serialize it to a json string.

Reply:
I have added string because we are reading data in old entity format in which `EntityId implements IdCompatible`. 5.0 onwards `EntityId` is no longer IdCompatible.

Reply:
I think what @chtyim is saying is that instead of doing:

```java
value = "{\n" +
            "  \"namespacedEntityId\": {\n" +
            "    \"type\": \"application\",\n" +
            "    \"id\": {\n" +
            "      \"namespace\": {\n" +
            "        \"id\": \"ns1\"\n" +
            "      },\n" +
            "      \"applicationId\": \"app1\"\n" +
            "    }\n" +
            "  },\n" +
            "  \"properties\": {\n" +
            "    \"akey1\" : \"avalue1\"\n" +
            "  },\n" +
            "  \"tags\": [\n" +
            "    \"cdap-data-pipeline\"\n" +
            "  ]\n" +
            "}";
```

You should do the following, because it is harder to mess up the quotes and other parts of the manual json string construction:
```java
    JsonObject namespaceId = new JsonObject();
    namespaceId.addProperty("id", "ns1");

    JsonObject applicationId = new JsonObject();
    applicationId.add("namespace", namespaceId);
    applicationId.addProperty("applicationId", "app1");

    JsonObject properties = new JsonObject();
    properties.addProperty("akey1", "avalue1");

    JsonArray tags = new JsonArray();
    tags.add(new JsonPrimitive("cdap-data-pipeline"));

    JsonObject jsonObject = new JsonObject();
    jsonObject.addProperty("type", "application");
    jsonObject.add("id",  applicationId);
    jsonObject.add("properties",  properties);
    jsonObject.add("tags",  tags);

    value = jsonObject.getAsString();
```